build/source/engine/computJacob.f90

```diff
@@ -0,0 +1,527 @@
+   ! ---------------------------------------------
+   if(computeVegFlux)then  ! (derivatives only defined when vegetation protrudes over the surface)
+  
+    ! liquid water fluxes for vegetation canopy (-)
+    aJac(ixVegWat,ixVegWat) = -scalarFracLiqVeg*(dCanopyEvaporation_dCanLiq - scalarCanopyLiqDeriv)*dt + 1._dp
+  
+    ! cross-derivative terms w.r.t. system temperatures (kg m-2 K-1)
+    aJac(ixVegWat,ixCasNrg) = -dCanopyEvaporation_dTCanair*dt
+    aJac(ixVegWat,ixVegNrg) = -dCanopyEvaporation_dTCanopy*dt + dt*scalarCanopyLiqDeriv*dCanLiq_dTcanopy
+    aJac(ixVegWat,ixTopNrg) = -dCanopyEvaporation_dTGround*dt
+  
+    ! cross-derivative terms w.r.t. canopy water (kg-1 m2)
+    aJac(ixTopWat,ixVegWat) = (dt/mLayerDepth(1))*(-scalarSoilControl*scalarFracLiqVeg*scalarCanopyLiqDeriv)/iden_water
+  
+    ! cross-derivative terms w.r.t. canopy temperature (K-1)
+    aJac(ixTopWat,ixVegNrg) = (dt/mLayerDepth(1))*(-scalarSoilControl*scalarCanopyLiqDeriv*dCanLiq_dTcanopy)/iden_water
+  
+    ! cross-derivative terms w.r.t. canopy liquid water (J m-1 kg-1)
+    ! NOTE: dIce/dLiq = (1 - scalarFracLiqVeg); dIce*LH_fus/canopyDepth = J m-3; dLiq = kg m-2
+    aJac(ixVegNrg,ixVegWat) = (dt/canopyDepth)   *(-dCanopyNetFlux_dCanLiq) - (1._dp - scalarFracLiqVeg)*LH_fus/canopyDepth   ! dF/dLiq
+    aJac(ixTopNrg,ixVegWat) = (dt/mLayerDepth(1))*(-dGroundNetFlux_dCanLiq)
+  
+    ! energy fluxes with the canopy air space (J m-3 K-1)
+    aJac(ixCasNrg,ixCasNrg) = (dt/canopyDepth)*(-dCanairNetFlux_dCanairTemp) + dMat(ixCasNrg)
+    aJac(ixCasNrg,ixVegNrg) = (dt/canopyDepth)*(-dCanairNetFlux_dCanopyTemp)
+    aJac(ixCasNrg,ixTopNrg) = (dt/canopyDepth)*(-dCanairNetFlux_dGroundTemp)
+  
+    ! energy fluxes with the vegetation canopy (J m-3 K-1)
+    aJac(ixVegNrg,ixCasNrg) = (dt/canopyDepth)*(-dCanopyNetFlux_dCanairTemp)
+    aJac(ixVegNrg,ixVegNrg) = (dt/canopyDepth)*(-dCanopyNetFlux_dCanopyTemp) + dMat(ixVegNrg)
+    aJac(ixVegNrg,ixTopNrg) = (dt/canopyDepth)*(-dCanopyNetFlux_dGroundTemp)
+  
+    ! energy fluxes with the surface (J m-3 K-1)
+    aJac(ixTopNrg,ixCasNrg) = (dt/mLayerDepth(1))*(-dGroundNetFlux_dCanairTemp)
+    aJac(ixTopNrg,ixVegNrg) = (dt/mLayerDepth(1))*(-dGroundNetFlux_dCanopyTemp)
+  
+   endif  ! if there is a need to compute energy fluxes within vegetation
+  
+   ! -----
+   ! * energy fluxes for the snow-soil domain...
+   ! -------------------------------------------
+   do iLayer=1,nLayers  ! loop through layers in the snow-soil domain
+    ! - define layer indices
+    jLayer = ixSnowSoilNrg(iLayer)
+    ! - compute the Jacobian
+    aJac(jLayer,jLayer)   = (dt/mLayerDepth(iLayer))*(-dNrgFlux_dTempBelow(iLayer-1) + dNrgFlux_dTempAbove(iLayer)) + dMat(jLayer)
+    if(iLayer > 1)       aJac(jLayer-nVarSnowSoil,jLayer) = (dt/mLayerDepth(iLayer-1))*( dNrgFlux_dTempBelow(iLayer-1) )
+    if(iLayer < nLayers) aJac(jLayer+nVarSnowSoil,jLayer) = (dt/mLayerDepth(iLayer+1))*(-dNrgFlux_dTempAbove(iLayer  ) )
```

Comment:
Yes, the logic is intentional. The derivative in the flux w.r.t. the layer above does not exist for the top layer. Similarly, the derivative in the flux w.r.t. the layer below does not exist for the bottom layer.